src/pj_init.c

```diff
@@ -27,320 +27,352 @@
+        if (pj_param_exists (start, next->param))
+            continue;
+
+        /* Don't default ellipse if datum, ellps or any ellipsoid information is set */
+        if (0==strncmp(next->param,"ellps=", 6)) {
+            if  (pj_param_exists (start, "datum"))  continue;
+            if  (pj_param_exists (start, "ellps"))  continue;
+            if  (pj_param_exists (start, "a"))      continue;
+            if  (pj_param_exists (start, "b"))      continue;
+            if  (pj_param_exists (start, "rf"))     continue;
+            if  (pj_param_exists (start, "f"))      continue;
+            if  (pj_param_exists (start, "e"))      continue;
+            if  (pj_param_exists (start, "es"))     continue;
+        }
 
-    if ( (fid = pj_open_lib(ctx,"proj_def.dat", "rt")) != NULL) {
-        next = get_opt(ctx, start, fid, "general", next, NULL);
-        pj_ctx_fseek(ctx, fid, 0, SEEK_SET);
-        next = get_opt(ctx, start, fid, name, next, NULL);
-        pj_ctx_fclose(ctx, fid);
+        /* If we're here, it's OK to append the current default item */
+        last = last->next = pj_mkparam(next->param);
     }
-    if (errno)
-        errno = 0; /* don't care if can't open file */
-    ctx->last_errno = 0;
+    last->next = 0;
 
-    return next;
+    pj_dealloc_params (ctx, defaults, 0);
+    return last;
 }
 
-/************************************************************************/
-/*                              get_init()                              */
-/************************************************************************/
-static paralist *get_init(projCtx ctx, paralist **start, paralist *next, char *name, int *found_def) {
-    char fname[MAX_PATH_FILENAME+ID_TAG_MAX+3], *opt;
-    PAFile fid;
-    paralist *init_items = NULL;
-    const paralist *orig_next = next;
+/*****************************************************************************/
+paralist *pj_expand_init(PJ_CONTEXT *ctx, paralist *init) {
+/******************************************************************************
+Append expansion of <key> to the paralist <init>. The expansion is appended,
+rather than inserted at <init>'s place, since <init> may contain
+overrides to the expansion. These must take precedence, and hence come first
+in the expanded list.
 
-    (void)strncpy(fname, name, sizeof(fname)-2);
-    fname[sizeof(fname)-2] = '\0';
+Consider e.g. the key 'foo:bar' which (hypothetically) expands to 'proj=utm
+zone=32 ellps=GRS80', i.e. a UTM projection on the GRS80 ellipsoid.
 
-    /*
-    ** Search for file/key pair in cache
-    */
+The expression 'init=foo:bar ellps=intl' will then expand to:
 
-    init_items = pj_search_initcache( name );
-    if( init_items != NULL )
-    {
-        next->next = init_items;
-        while( next->next != NULL )
-            next = next->next;
-        *found_def = 1;
-        return next;
-    }
+           'init=foo:bar ellps=intl proj=utm zone=32 ellps=GRS80',
```

Comment:
This is new behaviour (previously the overridden parameter would be left out) and is potentially very confusing. Not sure what I think about it. At the very least it should be documented somewhere.

Reply:
I considered changing init=foo to xnit=foo when expanding, and adding beginit=foo/endinit=foo sentinels around the expanded expression. This would ease the comprehension of list dumps. But since it is not necessary currently, I just left them out.

Leaving the init=foo thing in the list made it possible to handle lists with the init at the first position of the list, without having to juggle around with a pointer-to-pointer-to-list, which was confusing and error prone (perhaps partially because it - like everything else in that corner of the code - was undocumented).

So it made for a much cleaner call sequence, by giving up the possibility of dropping the list element. We can, however, still mark it as expanded, by changing init= to e.g. exit=, xnit=, tini=, or any other 4 letter tag, should it show necessary.

Since the overall design of the previous version of the code was highly confusing, and came with "plenty of rope to shoot oneself in the foot with", I went for a call convention that was as tightly pruned as at all possible.

Reply:
Actually now introduced the one line patch changinf an expanded init to an xnit

Reply:
Hmm.. Now that I read it again, I wasn't very clear in my comment. I have no problem with init=xxx staying in there - that's the way it has been previously. My issue was with having two occurences of `ellps` which makes it difficult to comprehend which one is used. But having though it over I don't think it is a problem since only the first occurence will get marked as "used".

I don't like the `xnit` variant of `init`.